README
This python library gets geo-tagged tweet count from a specified city.

Run: python infer-load.py 23 0

This will run for 23 minutes and start counting from the beginning.

Note if you rerun the command the "0" input for the redis dB will continue counting from last time.  "1" actually flushes
or resets the redis dB.

Make sure that Redis is installed from redis.io, python libraries redis, pyproj, TwitterAPI are installed.

Also be sure to make sure the redis server is running:
/etc/init.d/redis start

Software Architecture:

The software architecture consists of the following modules:

1. Google API: A module (function) for google API services that fetches bounding box in latitude/longitudes for a given address/city.

2. Lat-Long Manipulation: A module (class with several functions) for latitude-longitude manipulation. This has functionalities to convert between lat-long and UTM coordinates, create a fishnet grid (pixels) in an area based on input granularity, associate a given lat-long with a grid-pixel, convert between grid-pixel to center point in lat-long etc.

3. Twitter API: This is the most important module (few functions), that, continuosly tracks streams of tweets within a bounding box, tracks the count within each pixel in the fishnet grid, and saves the data into a persistent in-memory database called REDIS.

4. REDIS: REDIS is a very popular super-fast persistent in-memory (key, value) pair based DB that also periodically saves the data into disk. This module in the software initiates the connection to the redis server, initializes the DB if need be, and saves the data into the DB and also to an output CSV file in the end.




CHANGES.txt
Initial Repo for Twitter API Intensity Mapping Service 8-28-2014
Made modifications to the tile size from 100m to 5m
This version will search Denver as well


test.py
import TwitterAPI
import urllib, json

CK = 'YyxjZoC9JRIF2Ynlrdbi5sH9l'
CS = 'jFH6tuBHKaV6B0Am5cGCJpydZiYbSnuz6hqPlZZMSnzvM9jOF1'
AT = '2742390818-BYNy3F2ONHVe85nTby7UVm3JPuOn4DaGef2swip'
ATS = 'vicUnQoWftfkLhCW7wsJV1oNmmCAQG3j1EDvIuh3ss6GQ'
api = TwitterAPI.TwitterAPI(CK, CS, AT, ATS)
#r = api.request('statuses/filter', {'locations':'-74,40,-73,41'})
#for item in r.get_iterator():
#    print item

serviceurl = 'http://maps.googleapis.com/maps/api/geocode/json?'
import json
import urllib
address = 'Denver, CO'
url = serviceurl + urllib.urlencode({'sensor':'false', 'address': address})
uh = urllib.urlopen(url)
data = uh.read()
js = json.loads(str(data))

sw_lat = js['results'][0]['geometry']['bounds']['southwest']['lat']
sw_lng = js['results'][0]['geometry']['bounds']['southwest']['lng']
ne_lat = js['results'][0]['geometry']['bounds']['northeast']['lat']
ne_lng = js['results'][0]['geometry']['bounds']['northeast']['lng']
bbox_coods = [sw_lng, sw_lat, ne_lng, ne_lat]
bbox_strng = ','.join([str(x) for x in bbox_coods])

r = api.request('statuses/filter', {'locations': bbox_strng})
count = 0
tws = []
for item in r.get_iterator():
    try:
        tws.append(item['geo']['coordinates'])
        print (item['geo']['coordinates'])
    finally:
        print "No coordinates"
    count = count + 1
    if count == 20:
        break




save_data.py


import json, redis, csv

OUTPUT_FILE = 'output.json'
OUTPUT_CSV = 'denver.csv'
REDIS_SERVER = "localhost"


def write_as_json(tweet_count_dict):
	json_compatible_dict = ([{'location' : k, 'tweet_count' : tweet_count_dict[k]}
							for k in tweet_count_dict.keys()])
	with open(OUTPUT_FILE, 'w') as fw:
		json.dump(json_compatible_dict, fw)


def initiate_redis_DB(tweet_count_dict, reset_db):
	r_server = redis.Redis(REDIS_SERVER)
	# note this code only gets executed if the user input is 1 for reset
	if (reset_db == 1):
		r_server.flushdb()
		# Remove the grid as a dictionary primer
		if false:
			for key in tweet_count_dict:
				r_server.set(list(key), 0)
	return r_server

def save_redis_DB_to_disk(r_server):
	r_server.save()
	return


def save_raw_to_csv(tws):
	with open(OUTPUT_CSV, 'w') as fp:
		w = csv.writer(fp, delimiter=',')
		data = [[ "User_id", "latitude", "longitude", "time_stamp"]]
		for key in tws:
			data.append(key)
		w.writerows(data)
	return


def save_redis_DB_to_csv(r_server):
	with open(OUTPUT_CSV, 'w') as fp:
		w = csv.writer(fp, delimiter=',')
		#data = [[ "latitude", "longitude", "tweet_count"]]
		# With the change in dictionary r_server will consist of uid, location, time
		# each key will be a user
		data = [[ "User_id", "latitude", "longitude", "time_stamp"]]
		for key in r_server.keys():
			#longlat will NOT be the key in my modified dbase the user_id is tehe key and lat/long/time are values
			#longlat = json.loads(key) 
			user_id = json.loads(key)
			#old count that only added to grid points that had values.  recall we initialized the database with the entire grid.
			if False:
				count = json.loads(r_server.get(key))
				if (count > 0):
					new_data = [longlat[1], longlat[0], count] 
					data.append(new_data)
			if True:
				longlat = json.loads(r_server.get(key))
				if (longlat >0):
					print 'longlat', longlat, ' type =', type(longlat)
					new_data = [user_id, longlat[1], longlat[0], longlat[2]]
					data.append(new_data)
		w.writerows(data)
		return


infer_load.py
import ConfigParser, sys
import lat_long_calc as llc
import google_map_wrapper as gmw
import twitter_api_wrapper as taw
from save_data import *

CONFIG_FILENAME = 'load_inference_config.ini'

def get_location_info(config_filename):
	cfg = ConfigParser.ConfigParser()
	cfg.read(config_filename)
	city = cfg.get('LOCATION', 'CITY')
	state = cfg.get('LOCATION', 'STATE')
	address = ",".join([city, state])
	grid_length_mtrs = cfg.getfloat('LOCATION', 'GRID_LENGTH_IN_MTR')
	return address, grid_length_mtrs



def main(duration_minutes, reset_db):
	#First get the location information by calling the CONFIG PARSER
	address, grid_length_mtrs = get_location_info(CONFIG_FILENAME)
	print 'Address =', address, "  ", grid_length_mtrs,'Duration = ', duration_minutes
	#Use the Google serers to calculate the bounding box
	sw_longlat, ne_longlat, bbox_strng = gmw.get_bounding_box_longlats(address)
	print "SW long lat =", sw_longlat, 'NE long lat =', ne_longlat
	#Develop the Zone information
	zone = llc.zone_utm_longlat(sw_longlat, ne_longlat, grid_length_mtrs)
	# Only call count_tweets
	tweet_count_dict = taw.count_tweets(bbox_strng, zone, duration_minutes, reset_db)
	# The json file may just be an alternate format
	#write_as_json(tweet_count_dict)
	
	return tweet_count_dict


if __name__ == "__main__":
	if (len(sys.argv) < 2):
		print 'Enter duration of counting in minutes, Restart counting (0=No/1=Yes)?'
	else:
		main(int(sys.argv[1]), int(sys.argv[2]))

	




lat_long_calc.py
import pygmaps, pyproj, math


def get_utm_zone_number(longitude):
	return (int((longitude + 180)/6.0) % 60)


class zone_utm_longlat:
	def __init__(self, sw_longlat_tuple, ne_longlat_tuple, grid_length_mtrs):
		self.sw_longlat_tuple = sw_longlat_tuple
		self.ne_longlat_tuple = ne_longlat_tuple
		self.utm_zone = get_utm_zone_number(sw_longlat_tuple[0])
		self.geo_proj = pyproj.Proj(proj = 'utm', zone = self.utm_zone, ellps = 'WGS84')
		self.min_utm_cood_tuple = list(self.geo_proj(sw_longlat_tuple[0], sw_longlat_tuple[1]))
		self.grid_length_mtrs = grid_length_mtrs
		sw_grid_long, sw_grid_lat = self.geo_proj(self.min_utm_cood_tuple[0] + grid_length_mtrs,
													self.min_utm_cood_tuple[1] + grid_length_mtrs, 
													inverse = True)
		self.grid_length_long = sw_grid_long - sw_longlat_tuple[0]
		self.grid_length_lat = sw_grid_lat - sw_longlat_tuple[1]
		return


	def convert_latlong_to_utm(self, query_longlat_tuple):
		x_cood, y_cood = self.geo_proj(query_longlat_tuple[0], query_longlat_tuple[1])
		return x_cood, y_cood


	def convert_utm_to_longlat(self, query_utm_cood_tuple):
		longitude, latitude = self.geo_proj(query_utm_cood_tuple[0], 
											query_utm_cood_tuple[1], inverse = True)
		return longitude, latitude


	def get_grid_index(self, query_longlat_tuple):
		x_index = int((query_longlat_tuple[0] - self.sw_longlat_tuple[0])/(1.0 * self.grid_length_long))
		y_index = int((query_longlat_tuple[1] - self.sw_longlat_tuple[1])/(1.0 * self.grid_length_lat))
		return x_index, y_index


	def is_longlat_in_area(self, query_longlat_cood_tuple):
		if ((self.sw_longlat_tuple[0] <= query_longlat_cood_tuple[0] <= self.ne_longlat_tuple[0])
			and (self.sw_longlat_tuple[1] <= query_longlat_cood_tuple[1] <= self.ne_longlat_tuple[1])):
			return 1
		else:
			return 0




twitter_api_wrapper.py
import TwitterAPI
import urllib, json
from lat_long_calc import *
from datetime import *
import numpy as np
from save_data  import *
import redis


CK = 'YyxjZoC9JRIF2Ynlrdbi5sH9l'
CS = 'jFH6tuBHKaV6B0Am5cGCJpydZiYbSnuz6hqPlZZMSnzvM9jOF1'
AT = '2742390818-BYNy3F2ONHVe85nTby7UVm3JPuOn4DaGef2swip'
ATS = 'vicUnQoWftfkLhCW7wsJV1oNmmCAQG3j1EDvIuh3ss6GQ'

TIME_CHECK_INTERVAL_ITERATIONS = 5
REDIS_SAVE_EVERY_N_TWEETS = 100


def count_tweets(bbox_strng, zone, duration_minutes, reset_db):
	grid_index_dict, tweet_count_dict = create_count_dict(zone)
	# r_server is the redis read dictionary.
	# tweet_count_dict is overall bounding box of the city broken into grids.  
	# This allows the counts to be located within one of the tile points an counted.  
	r_server  = initiate_redis_DB(tweet_count_dict, reset_db)

	api = TwitterAPI.TwitterAPI(CK, CS, AT, ATS)
	end_time = datetime.now() + timedelta(seconds = int(60 * duration_minutes))
	r = api.request('statuses/filter', {'locations': bbox_strng})
	tws =[]
	count = 0
	
	for item in r.get_iterator():
		if (count % TIME_CHECK_INTERVAL_ITERATIONS == 0):
			if (datetime.now() > end_time):
				break

		if item['coordinates']:
			#twt_loc is already our exact coordinates
			twt_loc = item['coordinates']['coordinates']
			# We also need id and created_at
			twt_id = item['id']
			twt_time = item['created_at']
			data = [twt_id, twt_loc[1], twt_loc[0], twt_time]
			tws.append(data)
			# We will keep this code, but really not necessary since the twitter API filter on locations will only return tweets from our desired location.  The code below simply pegs one of the sub grids that were created
			if (zone.is_longlat_in_area(twt_loc) == 1):
				i, j = zone.get_grid_index(twt_loc)
				grid_key = grid_index_dict[(i, j)]
				tweet_count_dict[grid_key] += 1
				#tweet_count dictionary is incremented is later used
				# to count.  However REDIS is updated with the statement
				# below @ grid_key.
				# We should add a key in the r_server dictionary based
				# on the id of the tweeter.
				# Ripping out the grid as a key and replacing with ids.
				#r_server.incr(list(grid_key))
				r_server.append(twt_id,[twt_loc,twt_time])
		count += 1
		if (count % REDIS_SAVE_EVERY_N_TWEETS == 0):
			r_server.save()

	r_server.save()
	# The module call to save from the redis dB to the CSV is temporarily removed.
	#save_redis_DB_to_csv(r_server)
	# Instead we use a simple write of the data to csv.  please don't let Supratim see this.  He will be disappointed.
	save_raw_to_csv(tws)
	return tweet_count_dict


#The below routine would create raw tweet counts
def save_tweets(bbox_strng, max_tweets):
	api = TwitterAPI.TwitterAPI(CK, CS, AT, ATS)
	r = api.request('statuses/filter', {'locations': bbox_strng})
	count = 0
	tws = []
	for item in r.get_iterator():
		if (item['coordinates']):
			twt_loc = item['coordinates']['coordinates']
			# We also need id and created_at
			twt_id = item['id']
			twt_time = item['created_at']
			data = [twt_id, twt_loc, twt_time]
			tws.append(data)
			# older write of just the coordinateds.
			#tws.append(item['coordinates']['coordinates'])
		count = count + 1
		if count == max_tweets:
			break
	return tws



def create_count_dict(zone):
	tweet_count_dict = {}
	grid_index_dict = {}
	x_index_max, y_index_max = zone.get_grid_index(zone.ne_longlat_tuple)
	for i in range(0, x_index_max + 1):
		for j in range(0, y_index_max + 1):
			sw_grid_point_longlat = (zone.sw_longlat_tuple[0] + i * zone.grid_length_long, zone.sw_longlat_tuple[1] + j * zone.grid_length_lat)
			mid_point_grid = (sw_grid_point_longlat[0] + 0.5 * zone.grid_length_long, sw_grid_point_longlat[1] + 0.5 * zone.grid_length_lat)
			grid_index_dict[(i, j)] = mid_point_grid
			tweet_count_dict[mid_point_grid] = 0

	return grid_index_dict, tweet_count_dict
